File: leetcode/188.买卖股票的最佳时机 IV/188.买卖股票的最佳时机 IV.go
package main

func maxProfit(k int, prices []int) int {
	n := len(prices)
	if n == 0 {
		return 0
	}
	if k > n/2 {
		k = n / 2
	}
	dp := make([][][]int, n)
	for i := 0; i < n; i++ {
		kList := make([][]int, k+1)
		for j := 0; j < k+1; j++ {
			kList[j] = []int{0, -prices[i]}
		}
		dp[i] = kList
	}
	for i := 0; i < n; i++ {
		for j := k; j > 0; j-- {
			if i == 0 {
				continue
			}
			dp[i][j][0] = max(dp[i-1][j][0], dp[i-1][j][1]+prices[i])
			dp[i][j][1] = max(dp[i-1][j][1], dp[i-1][j-1][0]-prices[i])
		}
	}
	return dp[n-1][k][0]
}

func max(int1 int, int2 int) int {
	if int1 < int2 {
		return int2
	} else {
		return int1
	}
}

func main() {
	k := 2
	prices := []int{2, 4, 1}
	result := maxProfit(k, prices)
	println(result)
}

File: leetcode/188.买卖股票的最佳时机 IV/188.买卖股票的最佳时机 IV.py
from typing import List


class Solution:
    def maxProfit(self, k: int, prices: List[int]) -> int:
        if not prices:
            return 0
        n = len(prices)
        # 一次交易至少需要2天，n天只能做n/2次交易,k大于n/2也无用
        # 所以，k和n/2 取最小作为交易次数
        k = min(k, n//2)
        # 初始化dp数组
        dp = []
        for i in range(n):
            # 因为需要下标为k的数据，所以k+1才能得到
            kk = [[0, -prices[i]]] * (k+1)
            dp.append(kk)

        for i in range(n):
            for j in range(k, 0, -1):
                # 第0天情况已经赋值默认值，直接跳过
                if i == 0:
                    continue
                # 状态转移方程
                dp[i][j][0] = max(dp[i - 1][j][0], dp[i - 1][j][1] + prices[i])
                dp[i][j][1] = max(dp[i - 1][j][1], dp[i - 1][j - 1][0] - prices[i])
        return dp[n - 1][k][0]


if __name__ == '__main__':
    prices = [2, 4, 1]
    k = 2
    result = Solution().maxProfit(k, prices)
    print(result)
